notes/overriding.md

```diff
@@ -0,0 +1,957 @@
+mechanism:
+
+**order-dependency**
+
+Inheritance is order dependent: the chain of extensions must be defined with the
+precise order they will be applied in mind, and definitions must be grouped in
+consequence. On the other hand, merging is commutative, and the precedence
+information is encoded as priorities. Thus, overriding by merging can be defined
+using stand-alone pieces of data, although the behavior of priorities is not
+local.
+
+**(EXP)**
+
+Merged records have only access to the final computed fixpoint `self`, while
+objects have access to the previous stage of extension via `super`. However, as
+in the NixOS module system, it is possible to address this issue using custom
+merge functions. This is a bit less expressive, but in a good way: it forces the
+merge strategy to be uniform along each field, while mechanisms like overlays or
+inheritance can do pretty much anything.
+
+## Proposal overview
+
+We basically propose to adopt the same kind of merge, priority, and custom merge
+function-based scheme as the NixOS module system. One big difference with the
+NixOS module system is that it would be built-in in the language, being usable
+for any native record whatsoever, and having good interaction with the rest of
+the language features. The implementation would have first class access to
+locations, to the AST, the memory layout, the metadata, and so on.
+
+We first describe the ideas at a high-level. Then, we review the issue and
+challenges of this approach, and propose concrete solutions to overcome or
+mitigate them. Finally, the precise operational semantics is laid out in a
+dedicated section.
+
+### Recursive records & merging
+
+**This section defines the semantics, rather than an actual efficient
+implementation**.
+
+As before, it is useful to see recursive records represented as functions - or
+constructors - that take a `self` parameter and return a final non recursive record,
+in the same way as the original overriding mechanism of Nixpkgs (the non-existing
+syntax `def := value` is used to insist on the fact that we are defining new objects):
+
+```
+r = {
+  a = 1;
+  b = a + 1;
+}
+// Definition of the representation of r
+repr(r) := fun self => {
+  a = 1;
+  b = self.a + 1;
+}
+```
+
+Field access amounts to compute a fix-point:
+
+```
+repr(r).foo := let fix = repr(r) fix in fix.foo
+```
+
+Merging simply merges the underlying representations:
+
+```
+repr(r1) & repr(r2) := fun self => r1 self & r2 self
+```
+
+### Priorities
+
+Merge is fundamentally commutative. The problem is, not all fields should be
+treated the same: some are default values that ought to be overridden, some are
+high-priority values that ought to override. The solution is to use a priority
+system, encoding the precedence of each value while retaining commutativity
+(only the priority annotation counts, not the order in which values are merged).
+
+Priorities have drawbacks. One is non-locality: the final result depends on
```

Comment:
This reminds me of precedence of binary operators. Some languages[citation needed] support rational (as opposed to integral) precedences, so that you can always insert one more layer between two existing ones. I don't see a reason not to do so here, even if we expect most users to stick to integers.

Reply:
It makes even more sense as Nickel has currently only one type of numbers, and doesn't feature a separate integer type.